packages/dev/core/src/Misc/observable.ts

```diff
@@ -195,17 +195,7 @@ export class Observable<T> {
      * @param unregisterOnFirstCall defines if the observer as to be unregistered after the next notification
      * @returns the new observer created for the callback
      */
-    public add(
-        callback: (eventData: T, eventState: EventState) => void,
-        mask: number = -1,
-        insertFirst = false,
-        scope: any = null,
-        unregisterOnFirstCall = false
-    ): Nullable<Observer<T>> {
-        if (!callback) {
-            return null;
-        }
-
+    public add(callback: (eventData: T, eventState: EventState) => void, mask: number = -1, insertFirst = false, scope: any = null, unregisterOnFirstCall = false): Observer<T> {

```

Comment:
I'm afraid there's some old code that calls `add` with a null / undefined parameter, which would explain why we tested it, and it would be a breaking change to modify that...

However, we could also consider it a bug to call `add` with the wrong parameter...

Let's see what others think.

Reply:
I guess a solution would be to still check it but throw instead of returning null. 
it is a breaking change though. 

I can't think of a reason to pass null as a callback, but as @Popov72 says, this was added for a reason.

I might have a different solution. Or maybe 2.

The first - when adding an observer with null as callback, set it to be a no-op and return the observer. This way the observer does nothing, but ` add`  still returns an observer.
The second - leave it as is and run the check in the `notifyObservers` whether or not callback is defined. This is not my favorite solution, as this is an extra check in code that runs many time each frame.

Reply:
Or we can just respect a callback type instead of "rewriting" to Nullable:

https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsAysAhsaBeK7EgDcAUMQMYA26AztVAPIBG1EATgqwDwAqAfFAG9SAX1KUadJi3bpGFCD35CoKqGACucgJZk8AEz08AwugoVG6MgGsoEAB5ZEeugAo3EBMgAimdAC4obgAaW09UDCwA+CRwzAgASihsfjgAey09RIAfKER1M3jeFzJTc0srAO4TMwtrEIBbGgrc9XrGNiSoAFoARhCtRGlgADEtVmpgToAzUxYQ6jJUyAD8EE68sxD1RFYIAHMtCbZ6RFHx4GqKadmEysvam3tHZxazKAB+V6uAqTYORUExFUwN2wHUrEQgmEeDoqxIwNEonIqUGk1SzD+bB66wgAHcGBiZHIFAIoNRUvUID4MAEJqwBnsoMIivESMjUWTND0-L92GxOKTyZTqf4ycB6YhGczOujpBwegA6dAGNyJZKCJmssQoiacxgAJj8GyuuFlmNYiuVehcxtZQA

Reply:
being mainly a typescript issue, i quite like your solution. feels a bit cumbersome, BUT it really just needs a type-only fix.

Reply:
This is perfect for backcompat without harming perf, love it :-)

Reply:
Overloading with the following type is better than using generics

```ts
public add(callback?: null|undefined,mask?: number,insertFirst?:boolean,scope?: any, unregisterOnFirstCall?:boolean): null;

public add(callback: (eventData: T, eventState: EventState) => void,mask?: number,insertFirst?:boolean,scope?: any,unregisterOnFirstCall?:boolean): Observer<T>;

public add(callback?: ((eventData: T, eventState: EventState) => void)|null|undefined,mask?: number,insertFirst?:boolean,scope?: any,unregisterOnFirstCall?:boolean): Nullable<Observer<T>>;

public add(callback?: ((eventData: T, eventState: EventState) => void)|null|undefined,mask: number = -1,insertFirst = false,scope: any = null,unregisterOnFirstCall = false): Nullable<Observer<T>>{

}
```

Reply:
Can you update the PR with it, it will be easier to review ?

Reply:
I started a new PR to update it because I found that I couldn't update the PR